### DISCUSSION TITLE
Messenger worker does not start programatically

### DISCUSSION THREAD

**Discussion by arsrobota** [symfony/symfony#54111] [Q&A] [answered]
Hello,
I'm working with new Scheduler and my scheduled tasks need to be dynamic (I must be able to add new ones or update existing). Each task is based on an entity instance.
Whenever I persist a new object, worker is stopped through a Doctrine event listener:
```
namespace App\EventListener;

use App\Entity\Fire\MailQuery;
use Doctrine\Bundle\DoctrineBundle\Attribute\AsEntityListener;
use Doctrine\ORM\Event\PostPersistEventArgs;
use Doctrine\ORM\Events;
use Symfony\Bundle\FrameworkBundle\Console\Application;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Output\NullOutput;
use Symfony\Component\HttpKernel\KernelInterface;

#[AsEntityListener(event: Events::postPersist, method: 'postPersist', entity: MailQuery::class)]
class EmailSchedulerInsertListener
{
    public function __construct(private readonly KernelInterface $kernel) {}

    public function postPersist(MailQuery $mail, PostPersistEventArgs $event): void
    {
        $stopApp = new Application($this->kernel);
        $stopApp->setAutoExit(false);
        $input = new ArrayInput([
            'command' => 'messenger:stop-workers',
        ]);
        $output = new NullOutput();
        $stopApp->run($input, $output);
    }
}
```
And as soon as the WorkerStoppedEVent is dispatched, worker is started through a Messenger event listener:
```
namespace App\EventListener;

use Symfony\Component\EventDispatcher\Attribute\AsEventListener;
use Symfony\Component\HttpKernel\KernelInterface;
use Symfony\Component\Messenger\Event\WorkerStoppedEvent;

final class SchedulerUpdateListener
{
    public function __construct(private readonly KernelInterface $kernel) {}

    /**
     * Démarre le worker dès que le worker a été arrêté
     */
    #[AsEventListener(event: WorkerStoppedEvent::class)]
    public function onWorkerStoppedEvent(WorkerStoppedEvent $event): void
    {
        $event->getWorker()->run();
    }
}
```
My problem is that the worker is well stopped, but it's never relaunched (whereas the console does not give me the prompt). I tried other ways to make it start:
- using the Process component
```
$startProcess = Process::fromShellCommandline(\sprintf('php %s\\bin\\console messenger:consume -vv', $this->kernel->getProjectDir()));
        $startProcess->setTimeout(300);
        $startProcess->run();
        if (!$startProcess->isSuccessful()) {
            throw new ProcessFailedException($startProcess);
        }
        $this->logger->info('[Messenger] Les workers ont été démarrés dans le cadre d\'un redémarrage');
```
- using the Application component
```
$startApp = new Application($this->kernel);
        $startApp->setAutoExit(false);
        $input = new ArrayInput([
            'command' => 'messenger:consume',
            '-vv' => true,
        ]);
        $output = new BufferedOutput();
        $startApp->run($input, $output);
```
And in both case worker does not start and the console gives me the prompt.
I'm quite sure I forgot something or did things wrong but I can't figure out what. Can anybody bring light to me ?

Have a nice day

EDIT:
Here is my scheduler, in which I fetch the entities I need. Since the worker does not restart, created/updated/deleted entites are not taken into account.
```
#[AsSchedule(name: 'email')]
class SendEmailSchedule implements ScheduleProviderInterface
{
    private Schedule $schedule;

    public function __construct(private EntityManagerInterface $em) {}

    public function getSchedule(): Schedule
    {
        $mails = $this->em->getRepository(MailQuery::class)->findAllSchedulable();
        $messages = \array_map(function(MailQuery $mail) {
            return RecurringMessage::cron(\implode(' ', $mail->getSchedule()), new SendEmail($mail));
        }, $mails);

        return $this->schedule ??= (new Schedule())
            ->with(...$messages);
    }
}
```

**Comment by MatTheCat** [comment]
This seems like an [X/Y problem](https://en.wikipedia.org/wiki/XY_problem) to me: your use-case looks like you only need one recurring task which would then be handled by fetching the concerned entities and doing what needs to be done with them.

**Reply by arsrobota** [reply]
Hi Mathieu, thanks for your reply.
I indeed fetch all entities in my scheduler, but fetching is performed only once at worker start so further created/updated/deleted entities are never fetched. I updated my question to show the scheduler.

Cheers

**Comment by stof** [comment]
Trying to make your web process start the messenger consumer is not the way things are supposed to work, for a simple reason: your consumer process becomes a child process of the web process. When the web process ends, the child process will also end.

And if you use `$process->run()`, your parent PHP process will wait for the child process to exit (that's what `->run()` is doing), so you would be waiting for your consumer to stop (and it won't stop by itself)

**Reply by arsrobota** [reply]
Hi Stof, thank you for your clarifications.
So I suppose the best way is to stop/restart worker frequently with systemd or something similar ?

**Reply by stof** [reply, marked as answer]
No. The best way would be to implement your feature with a recurring message that would read the entities to figure out what should be done (and then either do that work or publish it with the messenger component), so that each processing of the schedule reads the configuration instead of doing it only at the start while your entities change over time.

**Reply by arsrobota** [reply]
OK, that sounds better, indeed. I'm going this way.
Thanks for your advice.

Best regards
